### ISSUE TITLE
refactor org and space resources to not take control of user permissions

### ISSUE THREAD

**Issue by jcarrothers-sap** [cloudfoundry-community/terraform-provider-cf#167]
Currently the `cloudfoundry_org` and `cloudfoundry_space` resources take full control of the user permissions for the org/space that they manage.  I would like to propose that we deprecate, or remove, this functionality on these resources and create new resources dedicated to managing user permissions on orgs and spaces.  

This would allow the user of terraform to manage orgs/spaces while allowing user permissions to optionally be managed manually.

**Comment by jcarrothers-sap** [comment]
@gberche-orange @janosbinder @psycofdj @ArthurHlt 

I would like your input on this idea before I proceed with implementing it.  I have need for the user management of the space resource to be optional, in some fashion, and I believe this would be the best option.

**Comment by janosbinder** [comment]
Hi @jcarrothers-sap I do not see any risks in our scenarios by refactoring the user management into a new resource.

**Comment by gberche-orange** [comment]
@jcarrothers-sap I agree with the proposal to make current space/orgs users support optional. Let me rephrase your proposal to make sure I understood you correctly:

> Currently the `cloudfoundry_org` and `cloudfoundry_space` resources take full control of the user permissions for the org/space that they manage

 i.e. the following CC API V2 resources gets controlled by terraform, respectively
* [/v2/users/:guid/spaces/:space_guid](http://apidocs.cloudfoundry.org/1.27.0/users/associate_space_with_the_user.html) for  `cloudfoundry_space.developers`
* [/v2/users/:guid/managed_spaces/:managed_space_guid](http://apidocs.cloudfoundry.org/1.27.0/users/associate_managed_space_with_the_user.html) for `cloudfoundry_space.managers`
* [/v2/users/:guid/managed_spaces/:managed_space_guid](http://apidocs.cloudfoundry.org/1.27.0/users/associate_space_with_the_user.html) for `cloudfoundry_space.auditors`

>  I have need for the user management of the space resource to be optional, in some fashion, and I believe this would be the best option.

e.g.  `cloudfoundry_space.developers` could now accept a `null` value in addition to an array of user guid strings. When the `null` value is specified then the `/v2/users/:guid/*spaces` resources won't be invoked by the `cloudfoundry_space` resource. 

>  I would like to propose that we deprecate, or remove, this functionality on these resources and create new resources dedicated to managing user permissions on orgs and spaces.

This could e.g. be supported by the new fine grain resources, each having `user_guid`  and `space_guid` attributes, such as:
* `cloudfoundry_user_managed_space`
* `cloudfoundry_user_audited_space`
* `cloudfoundry_user_space`

I understand the use case of managing space/org users manually outside terraform.

Have you indeed yet met the need to manage org/spaces in a first tfstate file, and associated space/org users in a distinct tfstate files (or alternatively in distinct tf configs) ?

Such fine grain syntax above while providing full control, might be too verbose to some other users. I'm not yet convinced we'd need to deprecate the existing `cloudfoundry_space|cloudfoundry_organization.developers|auditors|managers`. We'll also need to maintain some backward compatibility as we reach 1.0 version. So keeping the syntaxic sugar in `cloudfoundry_space|cloudfoundry_organization.developers|auditors|managers` along with new resource seems preferable in my opinion.

Besides, beware that the permission model in cloudfoundry is evolving towards custom roles with the perm project likely in the upcoming 12 months. See related https://github.com/cloudfoundry-incubator/perm/wiki/API#perm-api

I feel in terms of priority, we first need to stabilize the current 1.0 milestone, in particular https://github.com/mevansam/terraform-provider-cf/issues/156#issuecomment-429231947 and #150 #159 before merging new work in.

**Comment by jcarrothers-sap** [comment]
@gberche-orange 
I was thinking something more along the lines of two new resources dedicated to managing users, one for spaces and one for orgs that would replicate the current syntax, just in a stand-alone terraform resource.  These new resources would also replicate the current behaviour of this properties as defined on the `cloudfoundry_space` and `cloudfoundry_org` resources.

Examples:
```
resource "cloudfoundry_space_users" "s1_users" {
    space = "${cloudfoundry_space.s1.id}"
    org = "${cloudfoundry_org.o1.id}"
    managers = [
        "${cloudfoundry_user.tl.id}"
    ]
    developers = [
        "${cloudfoundry_user.tl.id}",
        "${cloudfoundry_user.dev1.id}",
        "${cloudfoundry_user.dev2.id}"
    ]
    auditors = [
        "${cloudfoundry_user.adr.id}",
        "${cloudfoundry_user.dev3.id}"
    ]
}

resource "cloudfoundry_org_users" "o1_users" {
    org = "${cloudfoundry_org.o1.id}"
    managers = [
        "${cloudfoundry_user.tl.id}"
    ]
    billing_managers = [
        "${cloudfoundry_user.tl.id}",
    ]
    auditors = [
        "${cloudfoundry_user.adr.id}",
        "${cloudfoundry_user.dev3.id}"
    ]
}
```

**Comment by gberche-orange** [comment]
@jcarrothers-sap I believe that with a low granularity resource and use of terraform modules, users would be able to achieve use-case that a `cloudfoundry_space_users` or `cloudfoundry_org_users`, as well as other use-cases, such as the following:

As a cloudfoundry user, I would like to be assigned as a space manager role to a space, and then manage space users on my own outside of terraform (space developers and space auditors).

I'm therefore suggesting to consider the following resources that should feel familial to CF CLI [unset/set-org-role](http://cli.cloudfoundry.org/en-US/cf/set-org-role.html) and [unset/set-space-role](http://cli.cloudfoundry.org/en-US/cf/set-space-role.html)

```hcl
resource "cloudfoundry_space_role" "space_role" {
  space_id="${cloudfoundry_space.s1.id}"
  user_id="${cloudfoundry_user.tl.id}"
  role="SpaceDeveloper" //SpaceManager or SpaceAuditor are also valid values
}

resource "cloudfoundry_org_role" "org_role" {
  org_id="${cloudfoundry_org.o1.id}"
  user_id="${cloudfoundry_user.tl.id}"
  role="OrgManager" //BillingManager or OrgAuditor are also valid values
}
```

example of using modules 

```hcl
module "space_members" {
  source = "./space_members"
  space_id = "${cloudfoundry_space.demo-basel.id}"
    managers = [
        "${cloudfoundry_user.tl.id}"
    ]
    developers = [
        "${cloudfoundry_user.tl.id}",
        "${cloudfoundry_user.dev1.id}",
        "${cloudfoundry_user.dev2.id}"
    ]
    auditors = [
        "${cloudfoundry_user.adr.id}",
        "${cloudfoundry_user.dev3.id}"
    ]
}
```

and module declaration:
```hcl
# ./space_members/space_members.tf
variable "space_id" {
  type="string"
}

variable "developers" {
  type="list"
  default=[ ]
}
variable "managers" {
  type="list"
  default=[ ]
}
variable "auditors" {
  type="list"
  default=[ ]
}

resource "cloudfoundry_space_role" "space_dev_role" {
  count="${length(var.developers)}"
  space_id="${var.space_id}"
  user_id="${element(var.developers, count.index)}"
  role="SpaceDeveloper"
}

resource "cloudfoundry_space_role" "space_manager_role" {
  count="${length(var.managers)}"
  space_id="${var.space_id}"
  user_id="${element(var.managers, count.index)}"
  role="SpaceManager"
}

resource "cloudfoundry_space_role" "space_auditor_role" {
  count="${length(var.auditors)}"
  space_id="${var.space_id}"
  user_id="${element(var.auditors, count.index)}"
  role="SpaceAuditor"
}
```

**Comment by bcampoli** [comment]
To our organization, not having this feature basically breaks the usefulness of the provider. We operate a large cloud-foundry deployment used by hundreds of devs. Our use case is wanting to be able to create orgs and manage application security groups, quotas, and service accesses using terraform --but letting the developpers and org/space managers deal with user permissions and user access. 

Currently we run our custom fork of the provider which is basically just the original provider with the user permission stuff commented out, which works well enough but we would much prefer to be on the official branch :)

**Comment by janosbinder** [comment]
Thank you for your feedback. We are very much open for collaboration and feel free to engage into a discussion about your improvement over Slack. 

We also welcome PRs against the official branch, which we review and discuss with you, how we all can benefit from it.

**Comment by ArthurHlt** [comment]
This has been implemented in #203 in this style: 

```hcl
resource "cloudfoundry_space_users" "s1_users" {
    space = "${cloudfoundry_space.s1.id}"
    org = "${cloudfoundry_org.o1.id}"
    managers = [
        "${cloudfoundry_user.tl.id}"
    ]
    developers = [
        "${cloudfoundry_user.tl.id}",
        "${cloudfoundry_user.dev1.id}",
        "${cloudfoundry_user.dev2.id}"
    ]
    auditors = [
        "${cloudfoundry_user.adr.id}",
        "${cloudfoundry_user.dev3.id}"
    ]
}

resource "cloudfoundry_org_users" "o1_users" {
    org = "${cloudfoundry_org.o1.id}"
    managers = [
        "${cloudfoundry_user.tl.id}"
    ]
    billing_managers = [
        "${cloudfoundry_user.tl.id}",
    ]
    auditors = [
        "${cloudfoundry_user.adr.id}",
        "${cloudfoundry_user.dev3.id}"
    ]
}
```

To have the same patterns of resource org and space
